### PR TITLE
Unlock express

### PR DESCRIPTION
**Pull request by julien51**
# Description

Adding a package `@unlock-protocol/unlock-express` that can be used in Express application to easily perform server-side locking!

Deployed a dummy application to http://unlock-express-demo.herokuapp.com/

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: packages/unlock-express/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# unlock-express
+
+An express plugin to add locks to express applicattions. This enables server-side locking of resources in order to make sure only actual  (paying) members can access/view them.
```
Comment (pi0neerpat) [inline packages/unlock-express/README.md:3]: Spelling

File: packages/unlock-express/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  baseAuthRedirectUri: `/my-callback`,
+
+  // Customize behavior on failure (usually when the user refused to purchase a membership!)
+  onFailure: async (req, res) => {
+    res.send('Please make sure you purchase the membership!')
+  },
+
+  // Customize the params on redirect URL: useful to identify the user for updateUserEthereumAddress and getUserEthereumAddress (optional)
+  optionalRedirectParams: async (req) => {
+    return {
```
Comment (pi0neerpat) [inline packages/unlock-express/README.md:43]: Should probably include an example for the referral address

File: packages/unlock-express/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  // Customize te RPC provider URLs
+  providers: {
+    1: 'provider URL',
+    4: 'provider URL',
+    100: 'provider URL',
+    137: 'provider URL',
+  }
+
+  // Customise the baseUrl for the application (do not include a path!)
+  baseUrl: 'https://myapp.tld:port'
```
Comment (pi0neerpat) [inline packages/unlock-express/README.md:57]: Can this be optional instead?

File: packages/unlock-express/example/app.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable no-unused-vars */
+const express = require('express')
+const cookieParser = require('cookie-parser')
+
+const configureUnlock = require('../index')
```
Comment (pi0neerpat) [inline packages/unlock-express/example/app.js:6]: The example should be self-contained if possible could this be moved to the current directory?

File: packages/unlock-express/example/app.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        },
+      }
+    },
+    // Required: Yields the current visitors ethereum address (required). (could look up in database based on user info... etc)
+    getUserEthereumAddress: async (req) => {
+      return req.cookies.userAddress
+    },
+    // Required: Saves the address for the current user. Signature + signede message is included for additional check. See ethers.utils.verifyMessage for details.
+    updateUserEthereumAddress: async (
+      req,
```
Comment (pi0neerpat) [inline packages/unlock-express/example/app.js:32]: Unused variables could be prefixed with a “_”

File: packages/unlock-express/src/keyExpirationFor.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Wrapper around keyExpirationFor
+ * @param {*} provider
+ * @param {*} lock
+ * @param {*} userAddress
+ * @returns
+ */
+async function keyExpirationFor(provider, lock, userAddress) {
+  return new Promise((resolve, reject) => {
+    const data = JSON.stringify({
```
Comment (pi0neerpat) [inline packages/unlock-express/src/keyExpirationFor.js:11]: This is some elite web3 super coder magic

File: packages/unlock-express/index.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+  if (!config.providers) {
+    // eslint-disable-next-line no-param-reassign
+    config.providers = {}
+  }
+
+  if (!config.providers[1]) {
+    // eslint-disable-next-line no-param-reassign
+    config.providers[1] =
+      'https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161'
```
Comment (pi0neerpat) [inline packages/unlock-express/index.js:23]: Are you sure you want to use your private infura keys here? If they exceed their limits, the calls will stop working and people will not be able to figure out why

**Review by pi0neerpat — COMMENTED** [review]
This is gonna be such a great developer experience!!!

File: packages/unlock-express/README.md (hunk 2 above)
Comment (julien51) [reply packages/unlock-express/README.md:43]: That's a good point!

File: packages/unlock-express/README.md (hunk 3 above)
Comment (julien51) [reply packages/unlock-express/README.md:57]: Oh, it is optional. Let me add this in the comments

File: packages/unlock-express/src/keyExpirationFor.js (hunk 6 above)
Comment (julien51) [reply packages/unlock-express/src/keyExpirationFor.js:11]: oh yeah :) Devs should not have to worry about this tho.

File: packages/unlock-express/index.js (hunk 7 above)
Comment (julien51) [reply packages/unlock-express/index.js:23]: We need to put a default config tho, don't we?

**Comment by clemsos** [comment]
Very cool 👍

**Pull request merged**